### PR TITLE
Add backend for auth0 provider

### PR DESCRIPTION
**Pull request by dyudyunov**

### CODE REVIEW

**Review by OPersian — APPROVED** [review]

File: social_core/backends/auth0.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+"""
+Auth0 implementation based on https://auth0.com/docs/quickstart/webapp/django/01-login
+"""
+from six.moves.urllib import request
+from jose import jwt
```
Comment (cmltaWt0) [inline social_core/backends/auth0.py:5]: This dependency should be added to the requirements in this package.

To do so you have to add `python-jose` to the project requirements.

https://github.com/raccoongang/social-core/blob/master/setup.py#L47
https://github.com/raccoongang/social-core/blob/master/requirements-base.txt

Comment (cmltaWt0) [reply social_core/backends/auth0.py:5]: Why do we ever need this library? It isn't enough to use `jwt`?
If it required to use this lib - add a link to the Auth0 documentation describing this intention.

**Review by cmltaWt0 — CHANGES_REQUESTED** [review]

File: social_core/backends/auth0.py (hunk 1 above)
Comment (dyudyunov) [reply social_core/backends/auth0.py:5]: I have tried to use `jwt` but got errors while decoding.
`Jose` is also used in [auth0 tutorial](https://auth0.com/docs/quickstart/webapp/django/01-login).

Comment (dyudyunov) [reply social_core/backends/auth0.py:5]: Another [topic](https://github.com/python-social-auth/social-core/pull/309) (this PR was taken as base for our backend). As written there: 

> in order to use any jwt-related backend, you need to install python-social-auth with the openidconnect extra:

`pip install social-auth-core[openidconnect]`
[requirements-openidconnect.txt](https://github.com/python-social-auth/social-core/blob/3ac1fb890cae460dc54abdef369cb9fa28409e53/requirements-openidconnect.txt) contains `python-jose`

Comment (dyudyunov) [reply social_core/backends/auth0.py:5]: In our version of social-core `openidconnect` doesn't require `python-jose`, so I'll add it to requirements-base.txt

Comment (cmltaWt0) [inline social_core/backends/auth0.py:5]: Add a blank line, please. pep8 note.

File: social_core/backends/auth0.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    def get_user_id(self, details, response):
+        """Return current user id."""
+        return details['user_id']
+
+    def get_user_details(self, response):
+        # Obtain JWT and the keys to validate the signature
+        idToken = response.get('id_token')
+        jwks = request.urlopen("https://" + self.setting('DOMAIN') + "/.well-known/jwks.json")
+        issuer = "https://" + self.setting('DOMAIN') + "/"
+        audience = self.setting('KEY') #CLIENT_ID
```
Comment (cmltaWt0) [inline social_core/backends/auth0.py]:
```suggestion
        audience = self.setting('KEY')  # CLIENT_ID
```

File: social_core/backends/auth0.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        payload = jwt.decode(idToken, jwks.read(), algorithms=['RS256'], audience=audience, issuer=issuer)
+
+        return {
+            'fullname': payload['name'],
+            'picture': payload['picture'],
+            'user_id': payload['sub'],
+            'email': payload['email'],
+            }
```
Comment (cmltaWt0) [inline social_core/backends/auth0.py]:
```suggestion
        }
```

File: social_core/backends/auth0.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        idToken = response.get('id_token')
+        jwks = request.urlopen("https://" + self.setting('DOMAIN') + "/.well-known/jwks.json")
+        issuer = "https://" + self.setting('DOMAIN') + "/"
+        audience = self.setting('KEY') #CLIENT_ID
+        payload = jwt.decode(idToken, jwks.read(), algorithms=['RS256'], audience=audience, issuer=issuer)
+
+        return {
+            'fullname': payload['name'],
+            'picture': payload['picture'],
+            'user_id': payload['sub'],
```
Comment (cmltaWt0) [inline social_core/backends/auth0.py:40]: Remove trailing spaces, please.

File: social_core/backends/auth0.py (hunk 1 above)
Comment (cmltaWt0) [inline social_core/backends/auth0.py:5]:
```suggestion
from jose import jwt

```

**Review by cmltaWt0 — CHANGES_REQUESTED** [review]

**Review by cmltaWt0 — APPROVED** [review]

**Pull request merged**